src/indexed_crate.rs

```diff
@@ -41,6 +41,166 @@ pub struct IndexedCrate<'a> {
     pub(crate) manually_inlined_builtin_traits: HashMap<Id, Item>,
 }
 
+/// Map a Key to a List (Vec) of values
+///
+/// It also has some nice operations for pushing a value to the list, or extending the list with
+/// many values.
+struct MapList<K, V>(HashMap<K, Vec<V>>);
+
+impl<K: std::cmp::Eq + std::hash::Hash, V> FromIterator<(K, V)> for MapList<K, V> {
+    #[inline]
+    fn from_iter<T: IntoIterator<Item = (K, V)>>(iter: T) -> Self {
+        let iter = iter.into_iter();
+        let (min, max) = iter.size_hint();
+        // Reserving this space doesn't give a measurable gain, but neither does it hurt so we are
+        // keeping it around
+        let mut map = Self::with_capacity(max.unwrap_or(min));
```

Comment:
If this doesn't produce measurable gain, I'd prefer to not include it for complexity and maintenance reasons.

Given that reserving space sometimes seems to make performance worse, I think it's a risk with larger downside than upside. We don't have enough test infrastructure set up to monitor for and prevent performance regressions, so we should minimize the risk they happen unnoticed.

I like the comment below that says "we tried reserving space and this was the result." Would you mind switching to something similar here?

Reply:
No problem!